### PR TITLE
Fix SVG Upload

### PR DESCRIPTION
**Pull request by thomasdax98**
## Previously:

Uploading an SVG made the API crash

Bildschirmfoto 2023-11-07 um 15 21 58

## Reason:

The svgContainsJavaScript check was moved to the new `FileValidationService` in https://github.com/vivid-planet/comet/pull/1279. The method was now executed in `fileFilter()` within the `DamUploadFileInterceptor`. However, at this moment `file.path` is undefined because the file is only saved to disk after `fileFilter()` is done. Meaning, in fileFilter(), we don't have access to the file content.

Bildschirmfoto 2023-11-07 um 15 23 00

## Now:

svgContainsJavaScript was moved back into its own `FileValidationInterceptor`. The upload works as expected

### CODE REVIEW

**Review by johnnyomair — COMMENTED** [review]
Have you tried achieving the validation using `file.buffer` instead of `file.path`, which should be the complete file in memory? 

image


https://expressjs.com/en/resources/middleware/multer.html

**Comment by nsams** [comment]
that was my fault, sorry.

Why not move the whole validation (using fileValidationService) to the second interceptor? Then the FileValidationService itself could stay as it is.

**Comment by nsams** [comment]
> Have you tried achieving the validation using file.buffer instead of file.path, which should be the complete file in memory?

multer can be configured with different storage backends, but we have a temporary file backend.

The multer typings do not (or can not) know the backend. Generally I dislike multer (after working only a little bit with it).

**Comment by thomasdax98** [comment]
> Have you tried achieving the validation using file.buffer instead of file.path, which should be the complete file in memory?

@johnnyomair I don't know how. `file.buffer` is also undefined in `fileFilter()`. That has probably something to do with the backend as @nsams commented

**Comment by johnnyomair** [comment]
@thomasdax98 sorry, I haven't thought about my suggestion long enough: We're not using an in-memory storage for file uploads (which wouldn't work for large files anyway), so it's not possible to do this validation in memory

**Comment by thomasdax98** [comment]
> Why not move the whole validation (using fileValidationService) to the second interceptor? Then the FileValidationService itself could stay as it is.

I forgot another option. I could call the fileValidationService in DamUploadFileInterceptor similar to how I did it here: https://github.com/vivid-planet/comet/pull/1235/commits/9d521e9941af1c94120487521b2f13036a260c24#diff-05b672fd11239f5186690b84570f5f43773a6c48b9660af26972e0387da71d21R85-R115

Moving all the validations to the second interceptor feels wrong to me, because when it's executed, the file was already saved to a temporary location. fileFilter() (apparently) works solely with the metadata before the file is saved to disk.

**Comment by johnnyomair** [comment]
> Why not move the whole validation (using fileValidationService) to the second interceptor? Then the FileValidationService itself could stay as it is.

Wouldn't this result in uploading large files to the temporary storage before validating it?

**Comment by thomasdax98** [comment]
> > Why not move the whole validation (using fileValidationService) to the second interceptor? Then the FileValidationService itself could stay as it is.
> 
> Wouldn't this result in uploading large files to the temporary storage before validating it?

This is probably still prevented by https://github.com/vivid-planet/comet/blob/8ed9698122b8bf9eb011672fd6253d7fc1fae71b/packages/api/cms-api/src/dam/files/dam-upload-file.interceptor.ts#L38-L40 but I'm not 100% sure in which order these functions/configs are applied

File: packages/api/cms-api/src/dam/files/file-validation-interceptor.ts [hunk 1]
```diff
@@ -0,0 +1,38 @@
+import { CallHandler, ExecutionContext, HttpException, Injectable, NestInterceptor } from "@nestjs/common";
+import fs from "fs";
+import { Observable, throwError } from "rxjs";
+import * as util from "util";
+
+import { svgContainsJavaScript } from "./files.utils";
+
+const readFile = util.promisify(fs.readFile);
+const unlinkFile = util.promisify(fs.unlink);
```
[the remark this answers is not in the document]
Comment (thomasdax98) [reply packages/api/cms-api/src/dam/files/file-validation-interceptor.ts]: Good catch, I changed it. I just copied the previous FileValidationInterceptor without thinking about it

**Comment by johnnyomair** [comment]
Maybe split validation into `validateFileMetadata` and `validateFileContents`?

**Comment by nsams** [comment]
> Wouldn't this result in uploading large files to the temporary storage before validating it?

I'm assuming we delete this temporary file.

Basically it's a performance optimisation with unknown impact vs. code organized on a contral place

I'm fine with both ways

**Comment by thomasdax98** [comment]
@nsams How about this solution?

File: packages/api/cms-api/src/dam/files/dam-upload-file.interceptor.ts [hunk 2]
```diff
@@ -52,8 +52,19 @@ export function DamUploadFileInterceptor(fieldName: string): Type<NestIntercepto
...
+        async intercept(context: ExecutionContext, next: CallHandler<unknown>): Promise<Observable<unknown>> {
+            const fileInterceptor = await this.fileInterceptor.intercept(context, next);
+
+            const ctx = context.switchToHttp();
+            const file = ctx.getRequest().file;
+
+            const error = await this.fileValidationService.validateFileContents(file);
+
+            if (error) {
+                return throwError(() => new HttpException("Rejected File Upload: SVG must not contain JavaScript", 422));
```
Comment (johnnyomair) [inline packages/api/cms-api/src/dam/files/dam-upload-file.interceptor.ts:64]: This error message is too specific. There may be other content violations in the future. The error should be defined in validateFileContents

Comment (thomasdax98) [reply packages/api/cms-api/src/dam/files/dam-upload-file.interceptor.ts]: True, I missed that

File: packages/api/cms-api/src/dam/files/file-validation.service.ts [hunk 3]
```diff
@@ -37,15 +47,31 @@ export class FileValidationService {
...
+
+    // https://github.com/expressjs/multer/blob/master/storage/disk.js#L54-L62
+    private async removeFile(file: FileUploadInterface) {
+        const path = file.path;
+
+        delete (file as Partial<FileUploadInterface>).destination;
+        delete (file as Partial<FileUploadInterface>).filename;
+        delete (file as Partial<FileUploadInterface>).path;
+
+        await unlink(path);
```
Comment (thomasdax98) [inline packages/api/cms-api/src/dam/files/file-validation.service.ts:75]: Don't delete in FileValidationService

Comment (thomasdax98) [reply packages/api/cms-api/src/dam/files/file-validation.service.ts]: @nsams FYI: I moved the `removeFile()` method to files.utils.ts and call it in the interceptor (not the FileValidationService) now

File: packages/api/cms-api/src/dam/files/files.utils.ts [hunk 4]
```diff
@@ -106,3 +107,14 @@ export const svgContainsJavaScript = (svg: string) => {
 
     return recursivelyFindJSInSvg(jsonObj);
 };
+
+export const removeFile = async (file: FileUploadInterface) => {
```
Comment (nsams) [inline packages/api/cms-api/src/dam/files/files.utils.ts:111]: I would name that `removeMulterTempFile` or so

**Review by nsams — APPROVED** [review]

**Pull request merged**